File: 581 47-smooth triangular numbers.jl
#!/usr/bin/julia

# Daniel "Trizen" Șuteu
# License: GPLv3
# Date: 02 May 2017
# Edit: 28 June 2019
# https://github.com/trizen

# https://projecteuler.net/problem=581

# Runtime: 4.153s.

using Primes

function p_581()

    smooth = 47
    limit  = 1109496723125       # A002072(π(47)) = A002072(15)

    prime_nums = primes(smooth)
    smooth_prod = prod(prime_nums)

    function is_smooth_over_prod(n,k)

        g = gcd(n, k)

        while (g > 1)
            while (rem(n, g) == 0)
                n = div(n, g)
            end
            n == 1 && return true
            g = gcd(n, k)
        end

        n == 1
    end

    smooth_nums = Int64[1]

    for p in prime_nums
        for n in smooth_nums
            if (n*p <= limit)
                append!(smooth_nums, n*p)
            end
        end
    end

    total = 0

    for n in (smooth_nums)
        if (is_smooth_over_prod(n+1, smooth_prod))
            total += n
        end
    end

    return total
end

println(p_581())

File: 581 47-smooth triangular numbers.pl
#!/usr/bin/perl

# Daniel "Trizen" Șuteu
# License: GPLv3
# Date: 01 May 2017
# https://github.com/trizen

# https://projecteuler.net/problem=581

# Runtime: 1 min.

=for comment

Maximum value of n in the p-smooth T(n), is given by: A002072(prime_count(p)).

For p=47, the maximum value of n is 1109496723125.

Using the same idea from problem 204, we generate 47-smooth numbers not exceeding this limit.
Then for each 47-smooth number n, we check n+1 to see if is also 47-smooth.

=cut

use 5.010;
use strict;
use warnings;

use ntheory qw(valuation primes);

my $smooth = 47;
my @primes = @{primes($smooth)};

sub is_smooth {
    my ($n) = @_;

    foreach my $p (@primes) {
        if ($n % $p == 0) {
            $n /= $p**valuation($n, $p);
            return 1 if $n == 1;
        }
    }

    $n == 1;
}

my %seen;
my $sum    = 0;
my @smooth = (1);

foreach my $p (@primes) {
    foreach my $n (@smooth) {
        if (!exists($seen{$n}) and is_smooth($n+1)) {
            undef $seen{$n};
            $sum += $n;
        }

        if ($n*$p <= 1109496723125) {
            push @smooth, $n*$p;
        }
    }
    say "$p -> $sum";
}

say "Sum: $sum";
